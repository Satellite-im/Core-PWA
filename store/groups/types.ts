export interface GroupMember {
  name: string
  photoHash: string
  status: string
  address: string
  peerId?: string
}

export interface Group {
  id: string
  address: string
  name: string
  admin: string
  creator: string
  members: GroupMember[]
  addresses: string[]
  openInvites: boolean
  encryptionKey: string
  lastUpdate: number
}

export interface GroupsState {
  all: Group[]
  inviteSubscription: number | null
  groupSubscriptions: string[]
}

export enum GroupsError {
  INVITE_ALREADY_SENT = 'errors.groups.request_already_sent',
  RECIPIENT_NOT_FOUND = 'errors.groups.recipient_not_found',
  USER_NOT_INITIALIZED = 'errors.groups.user_not_initialized',
}
