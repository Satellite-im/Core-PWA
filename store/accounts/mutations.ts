import { AccountsState, RegistrationStatus } from './types'
import { User } from '~/libraries/Iridium/users/types'

const mutations = {
  setPin(state: AccountsState, pin: string) {
    state.pin = pin
  },
  setStorePin(state: AccountsState, storePin: boolean) {
    state.storePin = storePin
  },
  setRegistry(state: AccountsState, registry: boolean) {
    state.registry = registry
  },
  setPinHash(state: AccountsState, pinHash: string) {
    state.pinHash = pinHash
  },
  unlock(state: AccountsState, pin: string) {
    state.locked = false
    state.pin = pin
  },
  lock(state: AccountsState) {
    state.locked = true
  },
  setEncryptedPhrase(state: AccountsState, encryptedPhrase: string) {
    state.encryptedPhrase = encryptedPhrase
  },
  setPhrase(state: AccountsState, decryptedPhrase: string) {
    state.phrase = decryptedPhrase
  },
  setActiveAccount(state: AccountsState, activeAccountPubkey: string) {
    state.active = activeAccountPubkey
    state.initialized = true
  },
  setProfilePicture(state: AccountsState, image: string) {
    if (state.details) {
      state.details.photoHash = image
    }
  },
  setUserDetails(state: AccountsState, details: User) {
    state.details = {
      did: details.did,
      name: details.name,
      status: details.status,
      photoHash: details.photoHash,
    }
    state.welcomePopup = true
  },
  dismissWelcome(state: AccountsState) {
    state.welcomePopup = false
  },
  setUserPeerId(state: AccountsState, peerId: string) {
    if (state.details) {
      state.details = {
        ...state.details,
        peerId,
      }
    }
  },
  setRegistrationStatus(
    state: AccountsState,
    registrationStates: RegistrationStatus,
  ) {
    state.registrationStatus = registrationStates
  },
  setLastVisited(state: AccountsState, lastVisited: string) {
    state.lastVisited = lastVisited
  },
  setAdapter(state: AccountsState, adapter: string) {
    state.adapter = adapter
  },
  setEntropy(state: AccountsState, entropyMessage: string) {
    state.entropyMessage = entropyMessage
  },
}

export default mutations
