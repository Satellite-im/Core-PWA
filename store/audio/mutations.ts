import { AudioState } from './types'

const mutations = {
  mute(state: AudioState) {
    state.muted = !state.muted
  },
  toggleMute(state: AudioState) {
    state.muted = !state.muted
  },
  deafen(state: AudioState) {
    const isDeafened = state.deafened
    state.previousVolume = isDeafened ? state.previousVolume : state.volume
    state.volume = isDeafened ? state.previousVolume : 0
    state.deafened = !isDeafened
  },
  setVolume(state: AudioState, volume: Number) {
    state.previousVolume = volume
    state.volume = volume
  },
  setSystemVolume(state: AudioState, systemVolume: Number) {
    state.sounds.system = systemVolume
  },
  setInputVolume(state: AudioState, inputVolume: Number) {
    state.inputVolume = inputVolume
  },
  setSoundLevel(
    state: AudioState,
    sound: {
      inboundMedia: Number
      outboundMedia: Number
      system: Number
    },
  ) {
    state.sounds = sound
  },
}

export default mutations
