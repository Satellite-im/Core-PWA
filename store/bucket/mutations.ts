import { BucketState } from './types'
import { FilSystem } from '~/libraries/Files/FilSystem'

const mutations = {
  bucketInitialized(state: BucketState, status: boolean) {
    state.initialized = status
  },
  updateFileSystem(state: BucketState, fs: FilSystem) {
    state.fileSystem = fs
  },
  openDirectory(state: BucketState, name: string) {
    state.fileSystem.openDirectory(name)
  },
  createDirectory(state: BucketState, name: string) {
    state.fileSystem.createDirectory(name)
  },
  goBackToDirectory(state: BucketState, name: string) {
    state.fileSystem.goBackToDirectory(name)
  },
  createFile(state: BucketState, name: string) {
    state.fileSystem.createFile(name)
  },
}

export default mutations
