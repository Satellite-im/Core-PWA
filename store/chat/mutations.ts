import { ChatState } from './types'
import { ReplyObj, ChatTextObj } from '~/types/chat/chat'
import { UploadDropItemType } from '~/types/files/file'

const mutations = {
  chatText(state: ChatState, req: ChatTextObj) {
    state.chatTexts = state.chatTexts.some((item) => item.userId === req.userId)
      ? state.chatTexts.map((item) => {
          if (item.userId === req.userId) {
            return { ...item, value: req.value }
          }
          return item
        })
      : state.chatTexts.concat(req)
  },
  setChatReply(state: ChatState, req: ReplyObj) {
    state.replies = state.replies.some((item) => item.replyId === req.replyId)
      ? state.replies.map((item) => {
          if (item.replyId === req.replyId) {
            return { ...item, value: req.value }
          }
          return item
        })
      : state.replies.concat(req)
  },
  addFile(
    state: ChatState,
    {
      file,
      address,
    }: {
      file: UploadDropItemType
      address: string
    },
  ) {
    state.files[address]
      ? state.files[address].push(file)
      : (state.files[address] = [file])
  },
  setFiles(
    state: ChatState,
    {
      files,
      address,
    }: {
      files: UploadDropItemType[]
      address: string
    },
  ) {
    state.files[address] = files
  },
  deleteFiles(state: ChatState, address: string) {
    delete state.files[address]
  },
  setCountError(state: ChatState, countError: Boolean) {
    state.countError = countError
  },
  setAlertNsfw(state: ChatState, alertNsfw: Boolean) {
    state.alertNsfw = alertNsfw
  },
  setShowFilePreview(state: ChatState, showFilePreview: Boolean) {
    state.showFilePreview = showFilePreview
  },
}

export default mutations
