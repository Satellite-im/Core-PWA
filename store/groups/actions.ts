import Vue from 'vue'
import { map, uniq } from 'lodash'
import { DataStateType } from '../dataState/types'
import { Group, GroupMember, GroupsError, GroupsState } from './types'
import SolanaManager from '~/libraries/Solana/SolanaManager/SolanaManager'
import { ActionsArguments } from '~/types/store/store'
import TextileManager from '~/libraries/Textile/TextileManager'
import { GroupChatManager } from '~/libraries/Textile/GroupChatManager'
import BlockchainClient from '~/libraries/BlockchainClient'

const getGroupChatManager = (): GroupChatManager => {
  const $TextileManager: TextileManager = Vue.prototype.$TextileManager

  if (!$TextileManager.groupChatManager?.isInitialized()) {
    throw new Error(GroupsError.TEXTILE_NOT_INITIALIZED)
  }
  return $TextileManager.groupChatManager
}

const getUserAccount = () => {
  try {
    return BlockchainClient.getInstance().account
  } catch (e) {
    throw new Error(GroupsError.USER_NOT_INITIALIZED)
  }
}

export default {
  async initialize({ dispatch, state }: ActionsArguments<GroupsState>) {
    await Promise.all([
      dispatch('fetchGroups'),
      dispatch('subscribeToGroupInvites'),
      dispatch('subscribeToGroupsUpdate'),
    ])
    await Promise.all(
      state.all.map(async (group) => {
        await dispatch(
          'textile/subscribeToGroup',
          { groupId: group.id },
          { root: true },
        )
      }),
    )
  },

  /**
   * @method createGroup
   * @description create new group chat
   * @example
   */
  async createGroup(
    { commit }: ActionsArguments<GroupsState>,
    { name }: { name: string },
  ) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const groupChatManager = getGroupChatManager()

    const groupId = await groupChatManager.createGroupConversation()

    await $BlockchainClient.createGroup(groupId, name)

    return groupId
  },

  /**
   * @method fetchGroups
   * @description fetch groups in which user participates
   * @example
   */
  async fetchGroups({ commit, dispatch }: ActionsArguments<GroupsState>) {
    commit(
      'dataState/setDataState',
      { key: 'groups', value: DataStateType.Loading },
      { root: true },
    )

    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const { address } = getUserAccount()

    const groups = await $BlockchainClient.getUserGroups(address)
    const groupsUsers = await $BlockchainClient.getGroupsUsers(
      map(groups, 'id'),
    )

    const addresses = uniq([].concat(...(<any>map(groupsUsers, 'users'))))
    const usersInfo = await $BlockchainClient.getUsersInfo(addresses)

    const payload = groups.map((group) => {
      const members = groupsUsers
        .find((it) => it.id === group.id)
        ?.users?.map((address) =>
          usersInfo.find((user) => user.address === address),
        )
        .filter(Boolean) as GroupMember[]

      return { ...group, members: members || [] }
    })
    commit('setGroups', payload)
    await dispatch('fetchGroupsLastUpdate')
    commit(
      'dataState/setDataState',
      { key: 'groups', value: DataStateType.Ready },
      { root: true },
    )
  },

  /**
   * @method sendGroupInvite
   * @description invites user into group chat
   * @param group {Group} target group
   * @param recipient {string} recipient address
   * @example
   */
  async sendGroupInvite(
    { dispatch }: ActionsArguments<GroupsState>,
    { group, recipient }: { group: Group; recipient: string },
  ) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    await $BlockchainClient.inviteToGroup(group.id, recipient)

    dispatch(
      'textile/sendGroupMessage',
      {
        groupId: group.id,
        message: `\`I added ${recipient} to the chat\``,
      },
      { root: true },
    )
  },

  /**
   * @method subscribeToGroupInvites
   * @description subscribe to group chat invites
   * @example
   */
  async subscribeToGroupInvites({
    state,
    dispatch,
    commit,
  }: ActionsArguments<GroupsState>) {
    if (state.inviteSubscription !== null) {
      await dispatch('unsubscribeFromGroupInvites')
    }

    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const id = $BlockchainClient.addGroupInviteListener((payload) => {
      commit('addGroup', { ...payload, members: [] })
      dispatch('fetchGroupMembers', payload.id)
      dispatch('subscribeToGroupUpdate', payload.id)
    })

    commit('setInviteSubscription', id)
  },

  /**
   * @method unsubscribeFromGroupInvites
   * @description remove group chat invites listener
   * @example
   */
  async unsubscribeFromGroupInvites({
    state,
    commit,
  }: ActionsArguments<GroupsState>) {
    if (state.inviteSubscription) {
      const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
      await $BlockchainClient.unsubscribeGroupInvite(state.inviteSubscription)
      commit('setInviteSubscription', null)
    }
  },

  /**
   * @method subscribeToGroupsUpdate
   * @description subscribe to multiple groups update
   * @example
   */
  async subscribeToGroupsUpdate({
    state,
    dispatch,
    commit,
  }: ActionsArguments<GroupsState>) {
    if (state.groupSubscriptions.length) {
      await dispatch('unsubscribeFromGroupsUpdate')
    }
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    const ids = await $BlockchainClient.addGroupsListener((payload) => {
      commit('updateGroup', payload)
      dispatch('fetchGroupMembers', payload.id)
    })

    commit('setGroupSubscriptions', ids)
  },

  /**
   * @method subscribeToGroupUpdate
   * @description subscribe to group update
   * @example
   */
  async subscribeToGroupUpdate(
    { state, dispatch, commit }: ActionsArguments<GroupsState>,
    groupId: string,
  ) {
    await dispatch('textile/subscribeToGroup', { groupId }, { root: true })
    if (!state.groupSubscriptions.includes(groupId)) {
      const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
      const id = await $BlockchainClient.addGroupListener(
        groupId,
        (payload) => {
          commit('updateGroup', payload)
          dispatch('fetchGroupMembers', payload.id)
        },
      )
      commit('addGroupSubscription', id)
    }
  },

  /**
   * @method unsubscribeFromGroupsUpdate
   * @description remove all groups update invites listener
   * @example
   */
  async unsubscribeFromGroupsUpdate({
    state,
    commit,
  }: ActionsArguments<GroupsState>) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
    if (state.groupSubscriptions.length) {
      await $BlockchainClient.removeGroupListeners(state.groupSubscriptions)
      commit('setGroupSubscriptions', [])
    }
  },

  /**
   * @method addGroup
   * @description fetch group data and add it to group list
   * @param groupId {string} group id to add
   * @example
   */
  async addGroup(
    { commit, state }: ActionsArguments<GroupsState>,
    groupId: string,
  ) {
    if (!state.all.find((group) => group.id === groupId)) {
      const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()
      const group = await $BlockchainClient.getGroupById(groupId)

      commit('addGroup', group)
    }
  },

  /**
   * @method fetchGroupMembers
   * @description fetch groups members by given group id
   * @example
   */
  async fetchGroupMembers(
    { commit }: ActionsArguments<GroupsState>,
    groupId: string,
  ) {
    const $BlockchainClient: BlockchainClient = BlockchainClient.getInstance()

    const addresses = await $BlockchainClient.getGroupUsers(groupId)
    const members = await $BlockchainClient.getUsersInfo(addresses)
    commit('setGroupMembers', { groupId, members })
  },

  /**
   * @method fetchGroupsLastUpdate
   * @description fetch groups last message time
   * @example
   */
  async fetchGroupsLastUpdate({
    commit,
    state,
  }: ActionsArguments<GroupsState>) {
    const groupChatManager = getGroupChatManager()

    const ids = map(state.all, 'id')
    const results = await Promise.all(
      ids.map((id) => groupChatManager.getGroupLastUpdate(id)),
    )
    const payload = ids.reduce(
      (prev, curr, i) => ({ ...prev, [curr]: results[i] }),
      {},
    )
    commit('setGroupsLastUpdate', payload)
  },
}

export const exportForTesting = {
  getGroupChatManager,
  getUserAccount,
}
