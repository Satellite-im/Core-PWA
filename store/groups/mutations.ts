import { GroupsState, Group, GroupMember } from './types'

const mutations = {
  setGroups(state: GroupsState, groups: Group[]) {
    state.all = groups
  },
  addGroup(state: GroupsState, group: Group) {
    state.all.push(group)
  },
  updateGroup(state: GroupsState, group: Group) {
    state.all = state.all.map((item) =>
      item.id !== group.id ? item : { ...item, ...group },
    )
  },
  setInviteSubscription(state: GroupsState, value: number | null) {
    state.inviteSubscription = value
  },
  setGroupSubscriptions(state: GroupsState, value: string[]) {
    state.groupSubscriptions = value
  },
  addGroupSubscription(state: GroupsState, value: string) {
    state.groupSubscriptions.push(value)
  },
  setGroupMembers(
    state: GroupsState,
    { groupId, members }: { groupId: string; members: GroupMember[] },
  ) {
    const group = state.all.find((g) => g.id === groupId)
    if (group) {
      group.members = members.sort((a, b) => a.name.localeCompare(b.name))
    }
  },
  setGroupsLastUpdate(state: GroupsState, payload: { [key: string]: number }) {
    state.all = state.all.map((group) => ({
      ...group,
      lastUpdate: payload[group.id],
    }))
  },
}

export default mutations
