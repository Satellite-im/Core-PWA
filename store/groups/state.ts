import { GroupsState } from '~/store/groups/types'

const InitialGroupsState: GroupsState = {
  all: [],
  subscriptionId: null,
}

export default () => InitialGroupsState
