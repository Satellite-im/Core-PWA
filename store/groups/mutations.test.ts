import * as mutations from '~/store/groups/mutations'

describe('mutations.default.updateGroup', () => {
  test('0', () => {
    const result: any = mutations.default.updateGroup(
      { all: { map: () => 'Pierre Edouard' } },
      { id: 'a85a8e6b-348b-4011-a1ec-1e78e9620782' },
    )
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = mutations.default.updateGroup(
      { all: { map: () => 'George' } },
      { id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9' },
    )
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = mutations.default.updateGroup(
      { all: { map: () => 'Jean-Philippe' } },
      { id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9' },
    )
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const result: any = mutations.default.updateGroup(
      { all: { map: () => 'Michael' } },
      { id: 'a85a8e6b-348b-4011-a1ec-1e78e9620782' },
    )
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const result: any = mutations.default.updateGroup(
      { all: { map: () => 'Edmond' } },
      { id: 'a85a8e6b-348b-4011-a1ec-1e78e9620782' },
    )
    expect(result).toMatchSnapshot()
  })

  test('5', () => {
    const result: any = mutations.default.updateGroup(
      { all: { map: () => '' } },
      { id: '' },
    )
    expect(result).toMatchSnapshot()
  })
})

describe('mutations.default.addGroup', () => {
  test('0', () => {
    const result: any = mutations.default.addGroup(
      { all: { push: () => 0 } },
      'c466a48309794261b64a4f02cfcc3d64',
    )
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = mutations.default.addGroup(
      { all: { push: () => 64 } },
      'bc23a9d531064583ace8f67dad60f6bb',
    )
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = mutations.default.addGroup(
      { all: { push: () => 0 } },
      'bc23a9d531064583ace8f67dad60f6bb',
    )
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const result: any = mutations.default.addGroup(
      { all: { push: () => 32 } },
      'da7588892',
    )
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const result: any = mutations.default.addGroup(
      { all: { push: () => 64 } },
      '12345',
    )
    expect(result).toMatchSnapshot()
  })

  test('5', () => {
    const result: any = mutations.default.addGroup(
      { all: { push: () => Infinity } },
      '',
    )
    expect(result).toMatchSnapshot()
  })
})

describe('mutations.default.setGroupMembersInfo', () => {
  test('0', () => {
    const result: any = mutations.default.setGroupMembersInfo(
      { all: { map: () => 'Anas' } },
      'c466a48309794261b64a4f02cfcc3d64',
    )
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = mutations.default.setGroupMembersInfo(
      { all: { map: () => 'Anas' } },
      'da7588892',
    )
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = mutations.default.setGroupMembersInfo(
      { all: { map: () => 'Pierre Edouard' } },
      'bc23a9d531064583ace8f67dad60f6bb',
    )
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const result: any = mutations.default.setGroupMembersInfo(
      { all: { map: () => 'Anas' } },
      '9876',
    )
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const result: any = mutations.default.setGroupMembersInfo(
      { all: { map: () => 'Pierre Edouard' } },
      '9876',
    )
    expect(result).toMatchSnapshot()
  })

  test('5', () => {
    const result: any = mutations.default.setGroupMembersInfo(
      { all: { map: () => '' } },
      '',
    )
    expect(result).toMatchSnapshot()
  })
})
