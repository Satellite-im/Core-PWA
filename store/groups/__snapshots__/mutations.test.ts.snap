// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations.default.addGroup 0 1`] = `undefined`;

exports[`mutations.default.addGroup 1 1`] = `undefined`;

exports[`mutations.default.addGroup 2 1`] = `undefined`;

exports[`mutations.default.addGroup 3 1`] = `undefined`;

exports[`mutations.default.addGroup 4 1`] = `undefined`;

exports[`mutations.default.addGroup 5 1`] = `undefined`;

exports[`mutations.default.setGroupMembersInfo 0 1`] = `undefined`;

exports[`mutations.default.setGroupMembersInfo 1 1`] = `undefined`;

exports[`mutations.default.setGroupMembersInfo 2 1`] = `undefined`;

exports[`mutations.default.setGroupMembersInfo 3 1`] = `undefined`;

exports[`mutations.default.setGroupMembersInfo 4 1`] = `undefined`;

exports[`mutations.default.setGroupMembersInfo 5 1`] = `undefined`;

exports[`mutations.default.updateGroup 0 1`] = `undefined`;

exports[`mutations.default.updateGroup 1 1`] = `undefined`;

exports[`mutations.default.updateGroup 2 1`] = `undefined`;

exports[`mutations.default.updateGroup 3 1`] = `undefined`;

exports[`mutations.default.updateGroup 4 1`] = `undefined`;

exports[`mutations.default.updateGroup 5 1`] = `undefined`;
